### PR TITLE
fs: add mkdirSync and rmdirSync

### PR DESCRIPTION
**Pull request by sonnyp**

### CODE REVIEW

File: packages/fs/index.js [hunk 1]
```diff
@@ -72,8 +72,31 @@ function readFileSync(path, options = { encoding: null, flag: 'r' }) {
   return data.toString(encoding);
 }
 
+// TODO: mode
+function mkdirSync(path, mode = 0o777) {
+  const file = Gio.File.new_for_path(path);
```
Comment (sonnyp) [inline packages/fs/index.js]: I wanted to use GLib for mkdir too but for some reason GLib.mkdir is undefined.

File: packages/fs/index.js [hunk 2]
```diff
@@ -72,8 +72,31 @@ function readFileSync(path, options = { encoding: null, flag: 'r' }) {
...
+
+  const ok = file.make_directory(null)
+
+  if (!ok) {
+    // TODO: throw a better error
+    throw new Error("failed to make directory");
```
Comment (WebReflection) [inline packages/fs/index.js]: `failed to make ${path} directory` ?

File: packages/fs/index.js [hunk 3]
```diff
@@ -72,8 +72,31 @@ function readFileSync(path, options = { encoding: null, flag: 'r' }) {
...
+  }
+}
+
+function rmdirSync(path) {
+  const result = GLib.rmdir(path)
+
+  if (result !== 0) {
+    // TODO: throw a better error
+    throw new Error("failed to remove directory");
```
Comment (WebReflection) [inline packages/fs/index.js]: `failed to remove ${path} directory` ?

File: packages/fs/index.js (hunk 1 above)
Comment (WebReflection) [inline packages/fs/index.js]: how about this? https://developer.gnome.org/glib/stable/glib-File-Utilities.html#g-mkdir

`GLib.mkdir(path, mode)`

File: packages/fs/index.js [hunk 4]
```diff
@@ -72,8 +72,31 @@ function readFileSync(path, options = { encoding: null, flag: 'r' }) {
   return data.toString(encoding);
 }
 
+// TODO: mode
+function mkdirSync(path, mode = 0o777) {
+  const file = Gio.File.new_for_path(path);
+
+  const ok = file.make_directory(null)
+
+  if (!ok) {
```
Comment (WebReflection) [inline packages/fs/index.js]: `if (GLib.mkdir(path, mode) < 0)`

**Review by WebReflection — CHANGES_REQUESTED** [review]
no need for Gio

File: packages/fs/index.js (hunk 1 above)
Comment (WebReflection) [reply packages/fs/index.js]: OK, I've read this only now ... which platform are you using ?

Comment (WebReflection) [reply packages/fs/index.js]: `mkdir_with_parents(pathname, mode)` is the exposed method: https://developer.gnome.org/glib/stable/glib-File-Utilities.html#g-mkdir-with-parents

It's the equivalent of `mkdir -p a/b/c` and I think, it's enough for CGJS

When in doubt, `cgjs-about gi.GLib`: https://github.com/cgjs/cgjs-about

**Comment by sonnyp** [comment]
@WebReflection :speech_balloon: addressed

I'm a bit worried about mkdir_with_parents, the behavior is quite different from Node.js

I already use cgjs-about, very handy :)

**Comment by WebReflection** [comment]
> I'm a bit worried about mkdir_with_parents, the behavior is quite different from Node.js

since that's the only method we have in GJS, you either file a bug in there and wait for it to be solved, or you use the exposed method that works regardless for everyone that knows the node API.

I am a **very** pragmatic person so guess my pick 😅

**Comment by WebReflection** [comment]
well done, my _either_ was about moving forward here, not about the need for that method for consistency sake.

But I'm pretty sure they'll answer YAGNI ... or maybe not 😄

**Comment by sonnyp** [comment]
https://gitlab.gnome.org/GNOME/gobject-introspection/merge_requests/16

**Comment by WebReflection** [comment]
btw, a proper way to implement async would be using Gio.File indeed. I cannot find anything related to mode in there though  ... how weird

**Pull request merged**